tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -48,134 +48,139 @@ class AutoScheduleGen : public Halide::Generator<AutoScheduleGen> {
     }
 
     void schedule() {
-        if (!auto_schedule) {
-            return;
+        if (auto_schedule) {
+            // To use the auto-scheduler, we need to provide estimates on all
+            // the input/output sizes including estimates on all the parameter
+            // values; otherwise, the auto-scheduler will throw an assertion.
+
+            // First, provide estimates (min and extent values) for each dimension
+            // of the input images ('input', 'filter', and 'bias') using the
+            // set_bounds_estimate() method. set_bounds_estimate() takes in
+            // (min, extent) of the corresponding dimension as arguments.
+            input.dim(0).set_bounds_estimate(0, 131);
+            input.dim(1).set_bounds_estimate(0, 131);
+            input.dim(2).set_bounds_estimate(0, 64);
+            input.dim(3).set_bounds_estimate(0, 4);
+
+            filter.dim(0).set_bounds_estimate(0, 3);
+            filter.dim(1).set_bounds_estimate(0, 3);
+            filter.dim(2).set_bounds_estimate(0, 64);
+            filter.dim(3).set_bounds_estimate(0, 64);
+
+            bias.dim(0).set_bounds_estimate(0, 64);
+
+            // Next, provide estimates on the parameter values using the
+            // set_estimate() method.
+            min_value.set_estimate(2.0f);
+
+            // Last, provide estimates (min and extent values) for each dimension
```

Comment:
Nit: you say "First, Next, Last", but I presume the order is unimportant and you can set the estimates in any order.